avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/StatementInfo.java

```diff
@@ -0,0 +1,122 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.avatica.jdbc;
+
+import com.google.common.base.Preconditions;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Objects;
+
+/**
+ *  All we know about a statement. Encapsulates a {@link ResultSet}.
+ */
+public class StatementInfo {
+  private static final String NEXT_METHOD_NAME = "next";
+
+  final Statement statement; // sometimes a PreparedStatement
+  private ResultSet resultSet; // a Proxy, not the real ResultSet.
+  private long position = 0;
+
+  /**
+   * Support a Proxy against ResultSet in which <code>position</code> is incremented after
+   * every call to <code>next()</code>.
+   */
+  private class ResultSetInvocationHandler implements InvocationHandler {
+    private final ResultSet realResultSet;
+
+    private ResultSetInvocationHandler(ResultSet resultSet) {
+      this.realResultSet = resultSet;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+      try {
+        Object o = method.invoke(realResultSet, args);
+        if (NEXT_METHOD_NAME.equals(method.getName())) {
+          position++;
+        }
+        return o;
+      } catch (InvocationTargetException ex) {
+        throw ex.getCause();
+      }
+    }
+
+  }
+
+  public StatementInfo(Statement statement) {
+    this.statement = Objects.requireNonNull(statement);
+  }
+
+  /**
+   * Set a ResultSet on this object.
+   * @param resultSet The current ResultSet
+   */
+  public void setResultSet(ResultSet resultSet) {
+    // Updates have null ResultSet from the Connection
+    if (null == resultSet) {
+      this.resultSet = null;
+      return;
+    }
+
+    Class<?> clz = resultSet.getClass();
+    // Create a Proxy around the ResultSet to let us increment `position` when `next()` is called
+    // but before the methods returns without re-implementing all of ResultSet.
+    this.resultSet = (ResultSet) Proxy.newProxyInstance(clz.getClassLoader(), clz.getInterfaces(),
+        new ResultSetInvocationHandler(Objects.requireNonNull(resultSet)));
+  }
+
+  /**
+   * @return The {@link ResultSet} for this Statement
+   */
+  public ResultSet getResultSet() {
```

Comment:
Clarify doc. Returns a proxy not a real result set, and throws NPE if not initialized.